x/storage/keeper/keeper_rate_limit.go

```diff
@@ -0,0 +1,232 @@
+	}
+	err = k.paymentKeeper.ApplyUserFlowsList(ctx, []paymenttypes.UserFlows{bill})
+	if err != nil {
+		return fmt.Errorf("apply user flows list failed: %s %w", bucketInfo.BucketName, err)
+	}
+	return nil
+}
+
+func getTotalOutFlowRate(flows []paymenttypes.OutFlow) sdkmath.Int {
+	totalFlowRate := sdkmath.ZeroInt()
+	for _, flow := range flows {
+		totalFlowRate = totalFlowRate.Add(flow.Rate)
+	}
+	return totalFlowRate
+}
+
+func (k Keeper) setZeroBucketFlowRateLimit(ctx sdk.Context, bucketInfo *types.BucketInfo, paymentAccount sdk.AccAddress, bucketName string) error {
+	currentRateLimit, found := k.getBucketFlowRateLimit(ctx, paymentAccount, bucketName)
+
+	// do nothing if the bucket flow rate limit is already 0
+	if found && currentRateLimit.FlowRateLimit.IsZero() {
+		return nil
+	}
+
+	internalBucketInfo := k.MustGetInternalBucketInfo(ctx, bucketInfo.Id)
+	// if the rate limit is not set and the payment account is not owned by the bucket owner,
+	// the net flow rate of the bucket should be zero, but it's fine to call `unChargeBucketReadStoreFee`, it will do nothing
+	return k.unChargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
+}
+
+func (k Keeper) setNonZeroBucketFlowRateLimit(ctx sdk.Context, bucketInfo *types.BucketInfo, paymentAccount sdk.AccAddress, bucketName string, rateLimit sdkmath.Int) error {
+	currentRateLimit, found := k.getBucketFlowRateLimit(ctx, paymentAccount, bucketName)
+
+	// do nothing if the bucket flow rate limit is already the same as the new rate limit
+	if found && currentRateLimit.FlowRateLimit.Equal(rateLimit) {
+		return nil
+	}
+
+	internalBucketInfo := k.MustGetInternalBucketInfo(ctx, bucketInfo.Id)
+
+	// if the rate limit is set to zero before, we should resume the payment of the bucket
+	if found && currentRateLimit.FlowRateLimit.IsZero() {
+		internalBucketInfo.PriceTime = ctx.BlockTime().Unix()
+
+		bill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+		if err != nil {
+			return fmt.Errorf("get bucket bill failed: %s %s", bucketInfo.BucketName, err.Error())
+		}
+		totalOutFlowRate := getTotalOutFlowRate(bill.Flows)
+		if totalOutFlowRate.GT(rateLimit) {
+			return fmt.Errorf("the total out flow rate(%s) of the bucket is greater than the new rate limit(%s)", totalOutFlowRate.String(), rateLimit.String())
+		}
+
+		return k.chargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
+	}
+
+	// if the rate limit is not set t0 zero before, we should check the net flow rate of the bucket
+	bill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+	if err != nil {
+		return fmt.Errorf("get bucket bill failed: %s %s", bucketInfo.BucketName, err.Error())
+	}
+	totalOutFlowRate := getTotalOutFlowRate(bill.Flows)
+	if totalOutFlowRate.GT(rateLimit) {
+		return fmt.Errorf("the total out flow rate(%s) of the bucket is greater than the new rate limit(%s)", totalOutFlowRate.String(), rateLimit.String())
```

Comment:
it is a bit not resonable  to not allow to set a ratelimit that below the current usage.